lib/elixir/unicode/security.ex

```diff
@@ -106,7 +106,133 @@ defmodule String.Tokenizer.Security do
+    #   confusability, as [unicode bidi algo] must be applied.
+    #   [...] a fast path can be used: [...] if X has no characters
+    #   w/bidi classes R or AL, bidiSkeleton(X) = skeleton(X)
+    #
+    if length(s) > 1 and any_rtl?(s) do
+      unbidify(s) |> Enum.map(&confusable_prototype/1)
+    else
+      Enum.map(s, &confusable_prototype/1)
+    end
+  end
+
+  # load direction-changing and neutral codepoints valid in idents
+  {rtls, neutrals} =
+    Path.join(__DIR__, "UnicodeData.txt")
+    |> File.read!()
+    |> String.split(["\r\n", "\n"], trim: true)
+    |> Enum.reduce({[], []}, fn line, {rtls, neutrals} = acc ->
+      with [point, _, _, _, bidi | _] <- String.split(line, ";"),
+           codepoint = String.to_integer(point, 16),
+           true <- String.Tokenizer.id_codepoint?(codepoint) do
+        # https://www.unicode.org/reports/tr44/tr44-32.html#Bidi_Class_Values
+        cond do
+          bidi in ~w(R AL)s -> {[codepoint | rtls], neutrals}
+          bidi in ~w(WS ON CS EN ES ET NSM)s -> {rtls, [codepoint | neutrals]}
+          true -> acc
+        end
+      else
+        _ -> acc
+      end
+    end)
+
+  rangify = fn [head | tail] ->
+    {first, last, acc} =
+      Enum.reduce(tail, {head, head, []}, fn
+        number, {first, last, acc} when number == first - 1 ->
+          {number, last, acc}
+
+        number, {first, last, acc} ->
+          {number, number, [{first, last} | acc]}
+      end)
+
+    [{first, last} | acc]
+  end
+
+  # direction of a codepoint. (rtl, neutral, ltr fallback)
+  for {first, last} <- rangify.(rtls) do
+    if first == last do
+      defp dir(unquote(first)), do: :rtl
+    else
+      defp dir(i) when i in unquote(first)..unquote(last), do: :rtl
+    end
+  end
+
+  for {first, last} <- rangify.(neutrals) do
+    if first == last do
+      defp dir(unquote(first)), do: :neutral
+    else
+      defp dir(i) when i in unquote(first)..unquote(last), do: :neutral
+    end
+  end
+
+  defp dir(i) when is_integer(i), do: :ltr
+  defp dir(_), do: {:error, :codepoint_must_be_integer}
+
+  defp any_rtl?(s), do: Enum.any?(s, &(:rtl == dir(&1)))
+
+  defp dir_compare(a, b) do
+    """
+    #{if any_rtl?(a), do: "\n\n" <> dir_breakdown(a)}
+    #{if any_rtl?(b), do: dir_breakdown(b)}
+    """
+  end
+
+  defp dir_breakdown(s) do
+    init = "'#{s}' includes right-to-left characters:\n"
+
+    for codepoint <- s, into: init do
+      hex = :io_lib.format(~c"~4.16.0B", [codepoint])
+      "  \\u#{hex} #{[codepoint]} #{dir(codepoint)}\n"
+    end
+  end
+
+  # make charlist match visual order by reversing spans of {rtl, neutral}
+  #  UTS39-28 4: '[...] if the strings are known not to contain explicit
+  #   directional formatting characters[...], the algorithm can
+  #   be drastically simplified, [...], obviating the need for
+  #   the [...] stack of the [unicode bidi algo]'
+  def unbidify([head | tail]), do: unbidify({tail, dir(head), [head], []})
```

Comment:
If we could have unit tests around this function, it would be very helpful!

Reply:
Fantastic -- will do that w/some more cases than the very limited ones in warning_test